lib/elixir/lib/code/normalizer.ex

```diff
@@ -0,0 +1,456 @@
+         [
+           {{:., dot_meta, [Kernel, :to_string]}, middle_meta, [middle]},
+           {:binary, binary_meta, context}
+         ]}
+
+      part ->
+        if state.escape_interpolation do
+          maybe_escape_literal(part, state)
+        else
+          part
+        end
+    end)
+  end
+
+  defp normalize_map_args([{key, value} | rest], state) do
+    key =
+      cond do
+        is_atom(key) and not module_atom?(key) ->
+          meta = [format: :keyword, line: state.parent_meta[:line]]
+          {:__block__, meta, [key]}
+
+        true ->
+          do_normalize(key, state)
+      end
+
+    value = do_normalize(value, state)
+
+    [{key, value} | normalize_map_args(rest, state)]
+  end
+
+  defp normalize_map_args([], _) do
+    []
+  end
+
+  defp normalize_kw_blocks(form, meta, args, state) do
+    {kw_blocks, leading_args} = List.pop_at(args, -1)
+
+    kw_blocks =
+      Enum.map(kw_blocks, fn {tag, block} ->
+        block = do_normalize(block, %{state | parent_meta: meta})
+
+        block =
+          case block do
+            {_, _, [[{:->, _, _} | _] = block]} -> block
+            block -> block
+          end
+
+        # Only wrap the tag if it isn't already wrapped
+        tag =
+          case tag do
+            {:__block__, _, _} -> tag
+            _ -> {:__block__, [line: meta[:line]], [tag]}
+          end
+
+        {tag, block}
+      end)
+
+    {_, _, [leading_args]} = do_normalize(leading_args, %{state | parent_meta: meta})
+
+    {form, meta, leading_args ++ [kw_blocks]}
+  end
+
+  defp normalize_list_elements(elems, state, keyword? \\ false)
+
+  defp normalize_list_elements([[{_, _, [{_, _}]}] = first | rest], state, keyword?) do
+    # Skip already normalized 2-tuples
+    [first | normalize_list_elements(rest, state, keyword?)]
+  end
+
+  defp normalize_list_elements([{left, right} | rest], state, keyword?) do
+    keyword? =
+      if not keyword? do
+        Enum.empty?(rest) or Inspect.List.keyword?(rest)
+      else
+        keyword?
+      end
+
+    module_atom? = module_atom?(left)
+
+    left = do_normalize(left, state)
+    right = do_normalize(right, state)
+
+    pair =
+      if keyword? and not module_atom? do
+        left = Macro.update_meta(left, &Keyword.put(&1, :format, :keyword))
+        {left, right}
+      else
+        {:__block__, [line: state.parent_meta[:line]], [{left, right}]}
+      end
+
+    [pair | normalize_list_elements(rest, state, keyword?)]
+  end
+
+  defp normalize_list_elements([first | rest], state, keyword?) do
+    [do_normalize(first, state) | normalize_list_elements(rest, state, keyword?)]
+  end
+
+  defp normalize_list_elements([], _state, _keyword?) do
+    []
+  end
+
+  defp maybe_escape_literal(string, %{escape: true}) when is_binary(string) do
+    if String.printable?(string) do
+      {string, _} = Code.Identifier.escape(string, ?")
+      IO.iodata_to_binary(string)
+    else
+      String.trim(inspect(string), "\"")
```

Comment:
Why do we need this trim? 🤔

Reply:
Actually it's not needed, I will remove it

Reply:
You also need to remember inspect has a limit of characters, so we need to disable any limit. Do we need to use inspect here? Could we always use Code.Identifier.escape instead? It is probably better to format "\u0000" as "\0" instead of a binary... especially because if we have a string, it was likely a string originally too.

Reply:
You're right, I got confused and tried to replicate `Macro.to_string` behavior and I did not consider that we want to keep them as strings to not change the semantics of the output. This should be fixed now :)